driver/pace/driver/grid.py

```diff
@@ -0,0 +1,273 @@
+
+    @property
+    def _namelist(self) -> Namelist:
+        return Namelist.from_f90nml(self._f90_namelist)
+
+    def _serializer(self, communicator: pace.util.CubedSphereCommunicator):
+        import serialbox
+
+        serializer = serialbox.Serializer(
+            serialbox.OpenModeKind.Read,
+            self.path,
+            "Generator_rank" + str(communicator.rank),
+        )
+        return serializer
+
+    def _get_serialized_grid(
+        self,
+        communicator: pace.util.CubedSphereCommunicator,
+        backend: str,
+    ) -> pace.stencils.testing.grid.Grid:
+        ser = self._serializer(communicator)
+        grid = TranslateGrid.new_from_serialized_data(
+            ser, communicator.rank, self._namelist.layout, backend
+        ).python_grid()
+        return grid
+
+    def get_grid(
+        self,
+        quantity_factory: QuantityFactory,
+        communicator: CubedSphereCommunicator,
+    ) -> Tuple[DampingCoefficients, DriverGridData, GridData]:
+
+        backend = quantity_factory.empty(
+            dims=[pace.util.X_DIM, pace.util.Y_DIM], units="unknown"
+        ).gt4py_backend
+
+        if self.serialized_grid:
+            logger.info("Using serialized grid data")
+            grid = self._get_serialized_grid(communicator, backend)
+            grid_data = grid.grid_data
+            driver_grid_data = grid.driver_grid_data
+            damping_coefficients = grid.damping_coefficients
+        else:
+            logger.info("Using a grid generated from metric terms")
+            grid = pace.stencils.testing.grid.Grid.with_data_from_namelist(
+                self._namelist, communicator, backend
+            )
+            metric_terms = pace.util.grid.MetricTerms(
+                quantity_factory=quantity_factory, communicator=communicator
+            )
+            grid_data = pace.util.grid.GridData.new_from_metric_terms(metric_terms)
+            damping_coefficients = DampingCoefficients.new_from_metric_terms(
+                metric_terms
+            )
+            driver_grid_data = pace.util.grid.DriverGridData.new_from_metric_terms(
+                metric_terms
+            )
+
+        return damping_coefficients, driver_grid_data, grid_data
+
+
+def _transform_horizontal_grid(self, metric_terms: MetricTerms) -> MetricTerms:
+    """
+    Uses the Schmidt transform to locally refine the horizontal grid.
+    """
+    grid = metric_terms.grid
+    lon_transform, lat_transform = direct_transform(
+        lon=grid.data[:, :, 0],
+        lat=grid.data[:, :, 1],
+        stretch_factor=self.stretch_factor,
+        lon_target=self.lon_target,
+        lat_target=self.lat_target,
+        np=grid.np,
+    )
+    grid.data[:, :, 0] = lon_transform[:]
+    grid.data[:, :, 1] = lat_transform[:]
+
+    metric_terms._grid.data[:] = grid.data[:]
+    metric_terms._init_agrid()
+
+    return metric_terms
+
+
+def _replace_vertical_grid(self, metric_terms) -> GridData:
```

Comment:
This only uses `ks` and `restart_path` from `self`, you should probably pass those directly instead.